### ISSUE TITLE
EventEmitter memory leak detected

### ISSUE THREAD

**Issue by ali1k** [yahoo/fluxible#125]
I receive the following warning when I put several React components on a page:

`(node) warning: possible EventEmitter memory leak detected. `
![screen shot 2015-04-07 at 21 57 43](https://cloud.githubusercontent.com/assets/2254157/7032185/4cf11a72-dd71-11e4-957a-714fac9933fd.png)

Could you please advise me how to resolve this issue?

**Comment by koulmomo** [comment]
This error usually occurs when you have 11+ components listening to the same store. Make sure to have fewer [controller-views](https://github.com/facebook/flux/issues/55) that do the actual store listening and pass store data as props to the rest of the components. 

Sometimes it's ok to have more controller-views listening to the same store (depends on the tradeoffs in your applications), in that case, after creating your store increase the max number of listeners for that store.

Though I'm not sure how you could do that since you would need access to the actual store instance (not the class) to call that method. see [EventEmitter docs](https://nodejs.org/api/events.html#events_emitter_setmaxlisteners_n)

**Comment by ali1k** [comment]
@koulmomo thank you for your answer. I will try to redesign my app with less store listeners. In my case it seems feasible though it makes the app logic more complex. Reading through stackoverflow I thought this is a bug in nodejs core: http://stackoverflow.com/questions/15581978/nodejs-how-to-debug-eventemitter-memory-leak-detected-11-listeners-added

As you suggested I might be able to solve it by increasing the number of listeners too:
'process.setMaxListeners(0);'

**Comment by felics** [comment]
@ali1k I got the same error message when starting out with flux. I ended up breaking my app into top-level, stateful components that host a number of "dumb" views that are populated via props and manipulate state via actions. This made my app far easier to reason about and provided less room for error and of course, less event listeners.

**Comment by koulmomo** [comment]
@ali1k here's a post explaining the [smart and dumb components](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0) pattern @felics mentioned.

btw `process.setMaxListeners(0);` would not solve your current problem.

if you update `dispatchr` to the latest version, it uses [`eventemmiter3`](https://github.com/yahoo/dispatchr/blob/360c99305a628af9ce23928ff4830091fbf2c240/package.json#L24) instead of node's `EventEmitter`, so should no longer [warn on exceeding max listeners](https://github.com/primus/EventEmitter3)

**Comment by ali1k** [comment]
@felics @koulmomo thank you for the link and hints. I removed my next level store and the issue is resolved. I am fine with the idea of smart and dumb components. However, this is a very simple model and for my case is not very efficient.
My parent component has a Web API request and sends the result as properties to sub-components. Based on the properties, sub-components need to request additional Web APIs in their own level. Therefore, I had created two stores, one for parent component and one for sub-components. I think if I handle all the requests via one store (for parent component), I have to re-render the whole UI each time I have a sub-request which is not efficient. Maybe you have a better idea to solve that issue.
